File: src/main/kotlin/com/iamalangreen/self/lifting/exercise.kt
package com.iamalangreen.self.lifting

import com.iamalangreen.self.Response
import com.iamalangreen.self.success
import io.hypersistence.utils.hibernate.type.array.ListArrayType
import jakarta.persistence.*
import org.hibernate.annotations.Type
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.stereotype.Service
import org.springframework.web.bind.annotation.*

data class ExerciseRequest(
    val id: Long?,
    val name: String,
    val description: String? = null,
    val mainMuscles: Set<Long> = setOf(),
    val supportMuscles: Set<Long> = setOf(),
    val cues: List<String> = listOf(),
)

data class ExerciseResponse(
    val id: Long,
    val name: String,
    val description: String? = null,
    val mainMuscles: Set<MuscleResponse> = setOf(),
    val supportMuscles: Set<MuscleResponse> = setOf(),
    val cues: List<String> = listOf(),
)

fun Exercise.toResponse(): ExerciseResponse {
    return ExerciseResponse(
        id!!,
        name,
        description,
        mainMuscles.map { it.toResponse() }.toSet(),
        supportMuscles.map { it.toResponse() }.toSet(),
        cues
    )
}

@RestController
@RequestMapping("/api/lifting/exercise")
class ExerciseController(val exerciseService: ExerciseService) {
    @PostMapping
    fun createExercise(@RequestBody request: ExerciseRequest): Response {
        val exercise = exerciseService.create(
            request.name,
            request.description,
            request.mainMuscles,
            request.supportMuscles,
            request.cues
        )
        return success(exercise.toResponse())
    }

    @PutMapping("/{id}")
    fun updateExercise(@PathVariable id: Long, @RequestBody request: ExerciseRequest): Response {
        val exercise = exerciseService.update(
            id,
            request.name,
            request.description,
            request.mainMuscles,
            request.supportMuscles,
            request.cues
        )
        return success(exercise.toResponse())
    }

    @GetMapping
    fun getAllExercises(): Response {
        return success(exerciseService.getAll().map { it.toResponse() })
    }
}

interface ExerciseService {
    fun create(
        name: String,
        description: String?,
        mainMuscleIds: Set<Long>,
        supportMuscleIds: Set<Long>,
        cues: List<String>
    ): Exercise
    fun update(
        id: Long,
        name: String,
        description: String?,
        mainMuscleIds: Set<Long>,
        supportMuscleIds: Set<Long>,
        cues: List<String>
    ): Exercise
    fun getById(id: Long): Exercise
    fun getAll(): List<Exercise>
}

@Service
class DefaultExerciseService(
    private val exerciseRepository: ExerciseRepository,
    private val muscleService: MuscleService,
) : ExerciseService {
    override fun create(
        name: String,
        description: String?,
        mainMuscleIds: Set<Long>,
        supportMuscleIds: Set<Long>,
        cues: List<String>
    ): Exercise {
        val mainMuscles = mainMuscleIds.map { muscleService.getById(it) }
        val supportMuscles = supportMuscleIds.map { muscleService.getById(it) }
        val exercise = Exercise(name = name)
        exercise.description = description
        exercise.mainMuscles.addAll(mainMuscles)
        exercise.supportMuscles.addAll(supportMuscles)
        exercise.cues.addAll(cues)
        return exerciseRepository.save(exercise)
    }

    override fun update(
        id: Long,
        name: String,
        description: String?,
        mainMuscleIds: Set<Long>,
        supportMuscleIds: Set<Long>,
        cues: List<String>
    ): Exercise {
        val exercise = getById(id)
        val mainMuscles = mainMuscleIds.map { muscleService.getById(it) }
        val supportMuscles = supportMuscleIds.map { muscleService.getById(it) }

        exercise.name = name
        exercise.description = description

        exercise.mainMuscles.clear()
        exercise.mainMuscles.addAll(mainMuscles)

        exercise.supportMuscles.clear()
        exercise.supportMuscles.addAll(supportMuscles)

        exercise.cues.clear()
        exercise.cues.addAll(cues)

        return exerciseRepository.save(exercise)
    }

    override fun getById(id: Long): Exercise {
        return exerciseRepository.findById(id).orElseThrow()
    }

    override fun getAll(): List<Exercise> {
        return exerciseRepository.findAll()
    }
}

interface ExerciseRepository : JpaRepository<Exercise, Long>

@Entity
@Table(name = "lifting_exercise")
data class Exercise(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,
    @Column
    var name: String,
    @Column
    var description: String? = null,
    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(
        name = "lifting_main_muscle_exercise",
        joinColumns = [JoinColumn(name = "exercise_id")],
        inverseJoinColumns = [JoinColumn(name = "muscle_id")]
    )
    var mainMuscles: MutableSet<Muscle> = mutableSetOf(),
    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(
        name = "lifting_support_muscle_exercise",
        joinColumns = [JoinColumn(name = "exercise_id")],
        inverseJoinColumns = [JoinColumn(name = "muscle_id")]
    )
    var supportMuscles: MutableSet<Muscle> = mutableSetOf(),
    @Type(ListArrayType::class)
    @Column(name = "cues", columnDefinition = "text[]")
    var cues: MutableList<String> = mutableListOf(),
) {

    override fun toString(): String {
        return "Exercise(name='$name', description=$description)"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Exercise

        return id == other.id
    }

    override fun hashCode(): Int {
        return id?.hashCode() ?: 0
    }
}

File: backend/app/api/exercise.py
from typing import Optional
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session
from app.core.database import get_db
from app.services.exercise import ExerciseService
from app.schemas.exercise import ExerciseRequest, ExerciseResponse
from app.schemas.muscle import MuscleResponse
from app.schemas.response import Response, PageResponse

router = APIRouter()


@router.post("", response_model=Response)
async def create_exercise(request: ExerciseRequest, db: Session = Depends(get_db)):
    """创建练习"""
    exercise = ExerciseService.create(
        db,
        name=request.name,
        description=request.description,
        main_muscle_ids=request.main_muscles,
        support_muscle_ids=request.support_muscles,
        cues=request.cues
    )
    
    # 构建响应
    main_muscles = [MuscleResponse.model_validate(muscle) for muscle in exercise.main_muscles]
    support_muscles = [MuscleResponse.model_validate(muscle) for muscle in exercise.support_muscles]
    cues = ExerciseService.get_cues_list(exercise)
    
    response_data = ExerciseResponse(
        id=exercise.id,
        name=exercise.name,
        description=exercise.description,
        main_muscles=main_muscles,
        support_muscles=support_muscles,
        cues=cues
    )
    
    return Response(data=response_data)


@router.put("/{exercise_id}", response_model=Response)
async def update_exercise(exercise_id: int, request: ExerciseRequest, db: Session = Depends(get_db)):
    """更新练习"""
    exercise = ExerciseService.update(
        db,
        exercise_id=exercise_id,
        name=request.name,
        description=request.description,
        main_muscle_ids=request.main_muscles,
        support_muscle_ids=request.support_muscles,
        cues=request.cues
    )
    
    # 构建响应
    main_muscles = [MuscleResponse.model_validate(muscle) for muscle in exercise.main_muscles]
    support_muscles = [MuscleResponse.model_validate(muscle) for muscle in exercise.support_muscles]
    cues = ExerciseService.get_cues_list(exercise)
    
    response_data = ExerciseResponse(
        id=exercise.id,
        name=exercise.name,
        description=exercise.description,
        main_muscles=main_muscles,
        support_muscles=support_muscles,
        cues=cues
    )
    
    return Response(data=response_data)


@router.get("", response_model=Response)
async def get_all_exercises(
    page: int = Query(0, ge=0, description="页码"),
    size: int = Query(20, ge=1, le=100, description="每页大小"),
    name: Optional[str] = Query(None, description="练习名称"),
    db: Session = Depends(get_db)
):
    """获取所有练习"""
    skip = page * size
    exercises = ExerciseService.get_all(db, name=name, skip=skip, limit=size)
    total = ExerciseService.get_total_count(db, name=name)
    
    items = []
    for exercise in exercises:
        main_muscles = [MuscleResponse.model_validate(muscle) for muscle in exercise.main_muscles]
        support_muscles = [MuscleResponse.model_validate(muscle) for muscle in exercise.support_muscles]
        cues = ExerciseService.get_cues_list(exercise)
        
        item = ExerciseResponse(
            id=exercise.id,
            name=exercise.name,
            description=exercise.description,
            main_muscles=main_muscles,
            support_muscles=support_muscles,
            cues=cues
        )
        items.append(item)
    
    page_response = PageResponse(
        total=total,
        page=page,
        size=size,
        items=items
    )
    
    return Response(data=page_response)


@router.get("/{exercise_id}", response_model=Response)
async def get_exercise(exercise_id: int, db: Session = Depends(get_db)):
    """根据ID获取练习"""
    exercise = ExerciseService.get_by_id(db, exercise_id)
    
    # 构建响应
    main_muscles = [MuscleResponse.model_validate(muscle) for muscle in exercise.main_muscles]
    support_muscles = [MuscleResponse.model_validate(muscle) for muscle in exercise.support_muscles]
    cues = ExerciseService.get_cues_list(exercise)
    
    response_data = ExerciseResponse(
        id=exercise.id,
        name=exercise.name,
        description=exercise.description,
        main_muscles=main_muscles,
        support_muscles=support_muscles,
        cues=cues
    )
    
    return Response(data=response_data)
